## 1. Importing the Data
<p><img src="https://assets.datacamp.com/production/project_1230/img/nyc.jpg" alt="New York City skyline" width="600px">
<br>
Welcome to New York City (NYC), one of the most-visited cities in the world. 
As a result, there are many <a href="https://www.airbnb.com/"><em>Airbnb</em></a> listings to meet the high demand for temporary lodging for anywhere between a few nights to many months. 
In this notebook, we will take a look at the NYC Airbnb market by combining data from multiple file types like <code>.csv</code>, <code>.tsv</code>, and <code>.xlsx</code>.</p>
<p><br><br>
We will be working with three datasets:</p>
<ol>
<li><p><code>"datasets/airbnb_price.csv"</code></p></li>
<li><p><code>"datasets/airbnb_room_type.xlsx"</code></p></li>
<li><p><code>"datasets/airbnb_last_review.tsv"</code></p></li>
</ol>
<p><br><br>
Our goals are to convert untidy data into appropriate formats to analyze, and answer key questions including:</p>
<ul>
<li>What is the average price, per night, of an Airbnb listing in NYC? </li>
<li>How does the average price of an Airbnb listing, per month, compare to the private rental market? </li>
<li>How many adverts are for private rooms? </li>
<li>How do Airbnb listing prices compare across the five NYC boroughs?</li>
</ul>

In [2]:
import pandas as pd
import numpy as np
import datetime as dt

# Load airbnb_price.csv, prices
prices = pd.read_csv("datasets/airbnb_price.csv")
prices

listing_id        price                    nbhood_full
0            2595  225 dollars             Manhattan, Midtown
1            3831   89 dollars         Brooklyn, Clinton Hill
2            5099  200 dollars         Manhattan, Murray Hill
3            5178   79 dollars      Manhattan, Hell's Kitchen
4            5238  150 dollars           Manhattan, Chinatown
5            5295  135 dollars     Manhattan, Upper West Side
6            5441   85 dollars      Manhattan, Hell's Kitchen
7            5803   89 dollars          Brooklyn, South Slope
8            6021   85 dollars     Manhattan, Upper West Side
9            6848  140 dollars         Brooklyn, Williamsburg
10           7097  215 dollars          Brooklyn, Fort Greene
11           7322  140 dollars             Manhattan, Chelsea
12           7726   99 dollars        Brooklyn, Crown Heights
13           8024  130 dollars           Brooklyn, Park Slope
14           8025   80 dollars           Brooklyn, Park Slope
15           8110  110 dollars           Brooklyn, Park Slope
16           8490  120 dollars   Brooklyn, Bedford-Stuyvesant
17           8505   60 dollars      Brooklyn, Windsor Terrace
18           9518   44 dollars              Manhattan, Inwood
19           9657  180 dollars        Manhattan, East Village
20           9668   50 dollars              Manhattan, Harlem
21           9704   52 dollars              Manhattan, Harlem
22           9782   55 dollars           Brooklyn, Greenpoint
23           9783   50 dollars              Manhattan, Harlem
24          10452   70 dollars   Brooklyn, Bedford-Stuyvesant
25          10962   89 dollars          Brooklyn, South Slope
26          11708   85 dollars             Brooklyn, Bushwick
27          12048   40 dollars     Manhattan, Lower East Side
28          12192   68 dollars        Manhattan, East Village
29          12299  120 dollars          Brooklyn, South Slope
...           ...          ...                            ...
25179    36233091   40 dollars             Brooklyn, Bushwick
25180    36235150  170 dollars         Brooklyn, Williamsburg
25181    36235154  138 dollars              Manhattan, Harlem
25182    36236013   75 dollars        Brooklyn, Cypress Hills
25183    36237529   65 dollars        Staten Island, Rosebank
25184    36238485   91 dollars         Brooklyn, Williamsburg
25185    36241283   50 dollars              Brooklyn, Midwood
25186    36241782  100 dollars         Brooklyn, Williamsburg
25187    36242888  220 dollars     Manhattan, Upper East Side
25188    36272851  135 dollars            Manhattan, Kips Bay
25189    36273046  120 dollars     Manhattan, Upper East Side
25190    36274037  170 dollars  Manhattan, Financial District
25191    36280646   16 dollars           Queens, Forest Hills
25192    36308600  150 dollars           Brooklyn, Park Slope
25193    36309284   45 dollars              Queens, Bayswater
25194    36310268   65 dollars                Queens, Maspeth
25195    36310855   68 dollars   Brooklyn, Bedford-Stuyvesant
25196    36318560  120 dollars             Manhattan, Midtown
25197    36335990   75 dollars  Manhattan, Financial District
25198    36344079  150 dollars                Queens, Astoria
25199    36351128  110 dollars     Manhattan, Upper West Side
25200    36351543   33 dollars   Brooklyn, Bedford-Stuyvesant
25201    36390226   45 dollars   Brooklyn, Bedford-Stuyvesant
25202    36411407  150 dollars     Manhattan, Lower East Side
25203    36413632  550 dollars     Brooklyn, Brooklyn Heights
25204    36425863  129 dollars     Manhattan, Upper East Side
25205    36427429   45 dollars               Queens, Flushing
25206    36438336  235 dollars     Staten Island, Great Kills
25207    36442252  100 dollars              Bronx, Mott Haven
25208    36455809   30 dollars             Brooklyn, Bushwick

[25209 rows x 3 columns]

In [3]:
# Load airbnb_room_type.xlsx, xls
xls = pd.ExcelFile("datasets/airbnb_room_type.xlsx")
xls

In [4]:
# Parse the first sheet from xls, room_types
room_types = xls.parse(0)
room_types

listing_id                                        description  \
0            2595                              Skylit Midtown Castle   
1            3831                    Cozy Entire Floor of Brownstone   
2            5099          Large Cozy 1 BR Apartment In Midtown East   
3            5178                    Large Furnished Room Near B'way   
4            5238                 Cute & Cozy Lower East Side 1 bdrm   
5            5295                   Beautiful 1br on Upper West Side   
6            5441                    Central Manhattan/near Broadway   
7            5803     Lovely Room 1, Garden, Best Area, Legal rental   
8            6021   Wonderful Guest Bedroom in Manhattan for SINGLES   
9            6848                   Only 2 stops to Manhattan studio   
10           7097                  Perfect for Your Parents + Garden   
11           7322                                    Chelsea Perfect   
12           7726    Hip Historic Brownstone Apartment with Backyard   
13           8024                     CBG CtyBGd HelpsHaiti rm#1:1-4   
14           8025                           CBG Helps Haiti Room#2.5   
15           8110                              CBG Helps Haiti Rm #2   
16           8490             MAISON DES SIRENES1,bohemian apartment   
17           8505                 Sunny Bedroom Across Prospect Park   
18           9518       SPACIOUS, LOVELY FURNISHED MANHATTAN BEDROOM   
19           9657                   Modern 1 BR / NYC / EAST VILLAGE   
20           9668                              front room/double bed   
21           9704                Spacious 1 bedroom in luxe building   
22           9782                  Loft in Williamsburg Area w/ Roof   
23           9783                                back room/bunk beds   
24          10452                              Large B&B Style rooms   
25          10962    Lovely room 2 & garden; Best area, Legal rental   
26          11708                          Cute apt in artist's home   
27          12048                LowerEastSide apt share shortterm 1   
28          12192                                ENJOY Downtown NYC!   
29          12299                Beautiful Sunny Park Slope Brooklyn   
...           ...                                                ...   
25179    36233091                                 1 room in bushwick   
25180    36235150    Sky View beautiful apt in Williamsburg New York   
25181    36235154                                   Top on Manhattan   
25182    36236013    READY BED GO the goal is to make u feel at home   
25183    36237529            Beautiful Cozy Room, 20mn to FREE ferry   
25184    36238485  Chic Room in Williamsburg + Backyard +Washer D...   
25185    36241283               Enjoy your friendly stay at my place   
25186    36241782  Classic Brooklyn Loft (East Williamsburg/Bushw...   
25187    36242888                 Upper East Site Luxury 2 Bed rooms   
25188    36272851    Private Bedroom in a Prewar Gramercy Building!*   
25189    36273046                  Elegant Spacious UES private room   
25190    36274037                            Cozy Wall Street Studio   
25191    36280646                       Cable and wfi, L/G included.   
25192    36308600          Cozy and Beautiful Park Slope One Bedroom   
25193    36309284  Private Room Near JFK, St John's Hospital, & B...   
25194    36310268                                     Full size room   
25195    36310855  Room 3: Cozy Room w Large Closet & Light Break...   
25196    36318560    Luxury Sun-filled Private Room near Time Square   
25197    36335990                Sunny Studio steps from Wall Street   
25198    36344079                       The Little Castle in Astoria   
25199    36351128   One bedroom without roomies, close to everything   
25200    36351543              2-MONTH SUBLEASE (WITH EARLY MOVE-IN)   
25201    36390226             Comfortable clean Bedstuy private room   
25202    36411407               Brand new 1 bed

In [5]:
# Load airbnb_last_review.tsv, reviews
reviews = pd.read_csv("datasets/airbnb_last_review.tsv", sep ="\t")
reviews

listing_id         host_name      last_review
0            2595          Jennifer      May 21 2019
1            3831       LisaRoxanne     July 05 2019
2            5099             Chris     June 22 2019
3            5178          Shunichi     June 24 2019
4            5238               Ben     June 09 2019
5            5295              Lena     June 22 2019
6            5441              Kate     June 23 2019
7            5803            Laurie     June 24 2019
8            6021           Claudio     July 05 2019
9            6848     Allen & Irina     June 29 2019
10           7097              Jane     June 28 2019
11           7322              Doti     July 01 2019
12           7726  Adam And Charity     June 22 2019
13           8024             Lisel     July 01 2019
14           8025             Lisel  January 01 2019
15           8110             Lisel     July 02 2019
16           8490          Nathalie     June 19 2019
17           8505           Gregory     June 23 2019
18           9518              Shon     June 15 2019
19           9657              Dana    April 19 2019
20           9668   Ssameer Or Trip     June 01 2019
21           9704              Teri     June 14 2019
22           9782            Andrea     June 15 2019
23           9783   Ssameer Or Trip     July 01 2019
24          10452            Angela      May 12 2019
25          10962            Laurie     June 21 2019
26          11708            Tyrome     June 22 2019
27          12048               Ben     July 05 2019
28          12192            Edward     June 21 2019
29          12299             Abdul      May 27 2019
...           ...               ...              ...
25179    36233091             Emily     July 03 2019
25180    36235150             Melki     July 07 2019
25181    36235154             Jason     July 03 2019
25182    36236013           Emanuel     July 05 2019
25183    36237529          Marouene     July 05 2019
25184    36238485              Ryan     July 04 2019
25185    36241283              Lana     July 02 2019
25186    36241782                Ap     July 06 2019
25187    36242888      Jean-Nicolas     July 03 2019
25188    36272851          Thikshan     July 07 2019
25189    36273046            Celine     July 06 2019
25190    36274037            Schmid     July 06 2019
25191    36280646             Chris     July 07 2019
25192    36308600            Kailey     July 05 2019
25193    36309284           Phoenix     July 04 2019
25194    36310268         Alejandro     July 06 2019
25195    36310855            Rehana     July 06 2019
25196    36318560         Linda Lou     July 06 2019
25197    36335990            Farina     July 03 2019
25198    36344079            Prince     July 04 2019
25199    36351128       David & Amy     July 05 2019
25200    36351543            Friday     July 06 2019
25201    36390226            Angela     July 08 2019
25202    36411407              Mike     July 06 2019
25203    36413632              Matt     July 07 2019
25204    36425863             Rusaa     July 07 2019
25205    36427429              H Ai     July 07 2019
25206    36438336               Ben     July 07 2019
25207    36442252            Blaine     July 07 2019
25208    36455809         Christine     July 08 2019

[25209 rows x 3 columns]

In [6]:
# Print the first five rows of each DataFrame
print(prices.head(), "\n", room_types.head(), "\n", reviews.head())

   listing_id        price                nbhood_full
0        2595  225 dollars         Manhattan, Midtown
1        3831   89 dollars     Brooklyn, Clinton Hill
2        5099  200 dollars     Manhattan, Murray Hill
3        5178   79 dollars  Manhattan, Hell's Kitchen
4        5238  150 dollars       Manhattan, Chinatown 
    listing_id                                description        room_type
0        2595                      Skylit Midtown Castle  Entire home/apt
1        3831            Cozy Entire Floor of Brownstone  Entire home/apt
2        5099  Large Cozy 1 BR Apartment In Midtown East  Entire home/apt
3        5178            Large Furnished Room Near B'way     private room
4        5238         Cute & Cozy Lower East Side 1 bdrm  Entire home/apt 
    listing_id    host_name   last_review
0        2595     Jennifer   May 21 2019
1        3831  LisaRoxanne  July 05 2019
2        5099        Chris  June 22 2019
3        5178     Shunichi  June 24 2019
4        5238          

## 2. Cleaning the price column
<p>Now the <code>DataFrames</code> have been loaded, the first step is to calculate the average price per listing by <code>room_type</code>. </p>
<p>You may have noticed that the <code>price</code> column in the <code>prices</code> DataFrame currently states each value as a string with the currency (dollars) following, i.e.,</p>
<pre><code>price
225 dollars
89 dollars
200 dollars</code></pre>
<p></p>
<p>We will need to clean the column in order to calculate the average price.</p>

In [7]:
# Remove whitespace and string characters from prices column
prices["price"] = prices["price"].str.replace(" dollars", "")
prices["price"]

0        225
1         89
2        200
3         79
4        150
5        135
6         85
7         89
8         85
9        140
10       215
11       140
12        99
13       130
14        80
15       110
16       120
17        60
18        44
19       180
20        50
21        52
22        55
23        50
24        70
25        89
26        85
27        40
28        68
29       120
        ... 
25179     40
25180    170
25181    138
25182     75
25183     65
25184     91
25185     50
25186    100
25187    220
25188    135
25189    120
25190    170
25191     16
25192    150
25193     45
25194     65
25195     68
25196    120
25197     75
25198    150
25199    110
25200     33
25201     45
25202    150
25203    550
25204    129
25205     45
25206    235
25207    100
25208     30
Name: price, Length: 25209, dtype: object

In [8]:
prices.info() # is a object dtype

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
listing_id     25209 non-null int64
price          25209 non-null object
nbhood_full    25209 non-null object
dtypes: int64(1), object(2)
memory usage: 590.9+ KB


In [9]:
# Convert prices column to numeric datatype
prices["price"] = pd.to_numeric(prices["price"])

prices.info()  # now the dtype for the price column is int64!

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
listing_id     25209 non-null int64
price          25209 non-null int64
nbhood_full    25209 non-null object
dtypes: int64(2), object(1)
memory usage: 590.9+ KB


In [10]:
# Print descriptive statistics for the price column
print(prices["price"].describe())

count    25209.000000
mean       141.777936
std        147.349137
min          0.000000
25%         69.000000
50%        105.000000
75%        175.000000
max       7500.000000
Name: price, dtype: float64


## 3. Calculating average price
<p>We can see three quarters of listings cost \$175 per night or less. </p>
<p>However, there are some outliers including a maximum price of \$7,500 per night! </p>
<p>Some of listings are actually showing as free. Let's remove these from the <code>DataFrame</code>, and calculate the average price.</p>

In [11]:
# Subset prices for listings costing $0, free_listings
free_listings = prices["price"] == 0
free_listings # returns boolean value 

# To view listings with free lisitngs
filtered_row = prices[free_listings]
filtered_row


listing_id  price                   nbhood_full
10861    20333471      0        Bronx, East Morrisania
10958    20523843      0            Brooklyn, Bushwick
11023    20639628      0  Brooklyn, Bedford-Stuyvesant
11024    20639792      0  Brooklyn, Bedford-Stuyvesant
11025    20639914      0  Brooklyn, Bedford-Stuyvesant
11577    21291569      0            Brooklyn, Bushwick
11589    21304320      0            Brooklyn, Bushwick

In [13]:
# Update prices by removing all free listings from prices
prices = prices.loc[~free_listings]


In [15]:
# Calculate the average price, avg_price
avg_price = round(prices["price"].mean(), 2)

141.82

In [16]:
# Print the average price
print("The average price per night for an Airbnb listing in NYC is ${}.".format(avg_price))

The average price per night for an Airbnb listing in NYC is $141.82.


## 4. Comparing costs to the private rental market
<p>Now we know how much a listing costs, on average, per night, but it would be useful to have a benchmark for comparison. 
According to <a href="https://www.zumper.com/rent-research">Zumper</a>, a 1 bedroom apartment in New York City costs, on average, $3,100 per month. Let's convert the per night prices of our listings into monthly costs, so we can compare to the private market. </p>

In [17]:
# Add a new column to the prices DataFrame, price_per_month
prices["price_per_month"] = prices["price"] * 365 / 12

In [19]:
# Calculate average_price_per_month
average_price_per_month = round(prices["price_per_month"].mean(), 2)

In [20]:
# Compare Airbnb and rental market
print("airbnb monthly costs are ${}, while in the private market you would pay {}.".format(average_price_per_month, "$3,100.00"))

airbnb monthly costs are $4313.61, while in the private market you would pay $3,100.00.


## 5. Cleaning the room type column
<p>Unsurprisingly, using Airbnb appears to be substantially more expensive than the private rental market. We should, however, consider that these Airbnb listings include single private rooms or even rooms to share, as well as entire homes/apartments. 
<br><br>
Let's dive deeper into the <code>room_type</code> column to find out the breakdown of listings by type of room. The <code>room_type</code> column has several variations for <code>private room</code> listings, specifically: </p>
<ul>
<li>"Private room"</li>
<li>"private room"</li>
<li>"PRIVATE ROOM"</li>
</ul>
<p>We can solve this by converting all string characters to lower case (upper case would also work just fine). </p>

In [23]:
# Convert the room_type column to lowercase
room_types["room_type"] = room_types["room_type"].str.lower()
room_types["room_type"].dtype

dtype('O')

In [25]:
# Update the room_type column to category data type
room_types["room_type"] = room_types["room_type"].astype("category")
room_types["room_type"].dtype

CategoricalDtype(categories=['entire home/apt', 'private room', 'shared room'], ordered=False)

In [26]:
# Create the variable room_frequencies
room_frequencies = room_types["room_type"].value_counts()

In [27]:
# Print room_frequencies
print(room_frequencies)

entire home/apt    13266
private room       11356
shared room          587
Name: room_type, dtype: int64


## 6. What timeframe are we working with?
<p>It seems there is a fairly similar sized market opportunity for both private rooms (45% of listings) and entire homes/apartments (52%) on the Airbnb platform in NYC.
<br><br></p>
<p>Now let's turn our attention to the <code>reviews</code> DataFrame. The <code>last_review</code> column contains the date of the last review in the format of "Month Day Year" e.g., May 21 2019. We've been asked to find out the earliest and latest review dates in the DataFrame, and ensure the format allows this analysis to be easily conducted going forwards. </p>

In [29]:
reviews["last_review"].dtype 

dtype('O')

In [33]:
# Change the data type of the last_review column to datetime
reviews["last_review"] = pd.to_datetime(reviews["last_review"])
reviews["last_review"].dtype

dtype('<M8[ns]')

In [34]:
# Create first_reviewed, the earliest review date
first_reviewed = reviews["last_review"].dt.date.min()
first_reviewed

datetime.date(2019, 1, 1)

In [36]:
# Create last_reviewed, the most recent review date
last_reviewed = reviews["last_review"].dt.date.max()
last_reviewed

datetime.date(2019, 7, 9)

In [37]:
# Print the oldest and newest reviews from the DataFrame
print("The latest Airbnb review is {}, the earliest review is {}".format(last_reviewed, first_reviewed))

The latest Airbnb review is 2019-07-09, the earliest review is 2019-01-01


## 7. Joining the DataFrames.
<p>Now we've extracted the information needed, we will merge the three DataFrames to make any future analysis easier to conduct. Once we have joined the data, we will remove any observations with missing values and check for duplicates.</p>

In [40]:
# Merge prices and room_types to create rooms_and_prices
rooms_and_prices = prices.merge(room_types, how="outer", on="listing_id")
rooms_and_prices

listing_id  price                    nbhood_full  price_per_month  \
0            2595  225.0             Manhattan, Midtown      6843.750000   
1            3831   89.0         Brooklyn, Clinton Hill      2707.083333   
2            5099  200.0         Manhattan, Murray Hill      6083.333333   
3            5178   79.0      Manhattan, Hell's Kitchen      2402.916667   
4            5238  150.0           Manhattan, Chinatown      4562.500000   
5            5295  135.0     Manhattan, Upper West Side      4106.250000   
6            5441   85.0      Manhattan, Hell's Kitchen      2585.416667   
7            5803   89.0          Brooklyn, South Slope      2707.083333   
8            6021   85.0     Manhattan, Upper West Side      2585.416667   
9            6848  140.0         Brooklyn, Williamsburg      4258.333333   
10           7097  215.0          Brooklyn, Fort Greene      6539.583333   
11           7322  140.0             Manhattan, Chelsea      4258.333333   
12           7726   99.0        Brooklyn, Crown Heights      3011.250000   
13           8024  130.0           Brooklyn, Park Slope      3954.166667   
14           8025   80.0           Brooklyn, Park Slope      2433.333333   
15           8110  110.0           Brooklyn, Park Slope      3345.833333   
16           8490  120.0   Brooklyn, Bedford-Stuyvesant      3650.000000   
17           8505   60.0      Brooklyn, Windsor Terrace      1825.000000   
18           9518   44.0              Manhattan, Inwood      1338.333333   
19           9657  180.0        Manhattan, East Village      5475.000000   
20           9668   50.0              Manhattan, Harlem      1520.833333   
21           9704   52.0              Manhattan, Harlem      1581.666667   
22           9782   55.0           Brooklyn, Greenpoint      1672.916667   
23           9783   50.0              Manhattan, Harlem      1520.833333   
24          10452   70.0   Brooklyn, Bedford-Stuyvesant      2129.166667   
25          10962   89.0          Brooklyn, South Slope      2707.083333   
26          11708   85.0             Brooklyn, Bushwick      2585.416667   
27          12048   40.0     Manhattan, Lower East Side      1216.666667   
28          12192   68.0        Manhattan, East Village      2068.333333   
29          12299  120.0          Brooklyn, South Slope      3650.000000   
...           ...    ...                            ...              ...   
25179    36241782  100.0         Brooklyn, Williamsburg      3041.666667   
25180    36242888  220.0     Manhattan, Upper East Side      6691.666667   
25181    36272851  135.0            Manhattan, Kips Bay      4106.250000   
25182    36273046  120.0     Manhattan, Upper East Side      3650.000000   
25183    36274037  170.0  Manhattan, Financial District      5170.833333   
25184    36280646   16.0           Queens, Forest Hills       486.666667   
25185    36308600  150.0           Brooklyn, Park Slope      4562.500000   
25186    36309284   45.0              Queens, Bayswater      1368.750000   
25187    36310268   65.0                Queens, Maspeth      1977.083333   
25188    36310855   68.0   Brooklyn, Bedford-Stuyvesant      2068.333333   
25189    36318560  120.0             Manhattan, Midtown      3650.000000   
25190    36335990   75.0  Manhattan, Financial District      2281.250000   
25191    36344079  150.0                Queens, Astoria      4562.500000   
25192    36351128  110.0     Manhattan, Upper West Side      3345.833333   
25193    36351543   33.0   Brooklyn, Bedford-Stuyvesant      1003.750000   
25194    36390226   45.0   Brooklyn, Bedford-Stuyvesant      1368.750000   
25195    36411407  150.0     Manhattan, Lower East Side      4562.500000   
25196    36413632  550.0     Brooklyn, Brooklyn Heights     16729.166667   
25197    36425863  129.0     Manhattan, Upper East Side      3923.750000   
25198    36427429   45.0               Queens, Flushing      1368.750000   
25199    36438336  235.0     Staten Island, Great Kills

In [44]:
# Merge rooms_and_prices with the reviews DataFrame to create airbnb_merged
airbnb_merged = rooms_and_prices.merge(reviews, how = "outer", on="listing_id")
airbnb_merged

listing_id  price                    nbhood_full  price_per_month  \
0            2595  225.0             Manhattan, Midtown      6843.750000   
1            3831   89.0         Brooklyn, Clinton Hill      2707.083333   
2            5099  200.0         Manhattan, Murray Hill      6083.333333   
3            5178   79.0      Manhattan, Hell's Kitchen      2402.916667   
4            5238  150.0           Manhattan, Chinatown      4562.500000   
5            5295  135.0     Manhattan, Upper West Side      4106.250000   
6            5441   85.0      Manhattan, Hell's Kitchen      2585.416667   
7            5803   89.0          Brooklyn, South Slope      2707.083333   
8            6021   85.0     Manhattan, Upper West Side      2585.416667   
9            6848  140.0         Brooklyn, Williamsburg      4258.333333   
10           7097  215.0          Brooklyn, Fort Greene      6539.583333   
11           7322  140.0             Manhattan, Chelsea      4258.333333   
12           7726   99.0        Brooklyn, Crown Heights      3011.250000   
13           8024  130.0           Brooklyn, Park Slope      3954.166667   
14           8025   80.0           Brooklyn, Park Slope      2433.333333   
15           8110  110.0           Brooklyn, Park Slope      3345.833333   
16           8490  120.0   Brooklyn, Bedford-Stuyvesant      3650.000000   
17           8505   60.0      Brooklyn, Windsor Terrace      1825.000000   
18           9518   44.0              Manhattan, Inwood      1338.333333   
19           9657  180.0        Manhattan, East Village      5475.000000   
20           9668   50.0              Manhattan, Harlem      1520.833333   
21           9704   52.0              Manhattan, Harlem      1581.666667   
22           9782   55.0           Brooklyn, Greenpoint      1672.916667   
23           9783   50.0              Manhattan, Harlem      1520.833333   
24          10452   70.0   Brooklyn, Bedford-Stuyvesant      2129.166667   
25          10962   89.0          Brooklyn, South Slope      2707.083333   
26          11708   85.0             Brooklyn, Bushwick      2585.416667   
27          12048   40.0     Manhattan, Lower East Side      1216.666667   
28          12192   68.0        Manhattan, East Village      2068.333333   
29          12299  120.0          Brooklyn, South Slope      3650.000000   
...           ...    ...                            ...              ...   
25179    36241782  100.0         Brooklyn, Williamsburg      3041.666667   
25180    36242888  220.0     Manhattan, Upper East Side      6691.666667   
25181    36272851  135.0            Manhattan, Kips Bay      4106.250000   
25182    36273046  120.0     Manhattan, Upper East Side      3650.000000   
25183    36274037  170.0  Manhattan, Financial District      5170.833333   
25184    36280646   16.0           Queens, Forest Hills       486.666667   
25185    36308600  150.0           Brooklyn, Park Slope      4562.500000   
25186    36309284   45.0              Queens, Bayswater      1368.750000   
25187    36310268   65.0                Queens, Maspeth      1977.083333   
25188    36310855   68.0   Brooklyn, Bedford-Stuyvesant      2068.333333   
25189    36318560  120.0             Manhattan, Midtown      3650.000000   
25190    36335990   75.0  Manhattan, Financial District      2281.250000   
25191    36344079  150.0                Queens, Astoria      4562.500000   
25192    36351128  110.0     Manhattan, Upper West Side      3345.833333   
25193    36351543   33.0   Brooklyn, Bedford-Stuyvesant      1003.750000   
25194    36390226   45.0   Brooklyn, Bedford-Stuyvesant      1368.750000   
25195    36411407  150.0     Manhattan, Lower East Side      4562.500000   
25196    36413632  550.0     Brooklyn, Brooklyn Heights     16729.166667   
25197    36425863  129.0     Manhattan, Upper East Side      3923.750000   
25198    36427429   45.0               Queens, Flushing      1368.750000   
25199    36438336  235.0     Staten Island, Great Kills

In [48]:
airbnb_merged.isna().sum()

listing_id          0
price               7
nbhood_full         7
price_per_month     7
description        10
room_type           0
host_name           8
last_review         0
dtype: int64

In [49]:
# Drop missing values from airbnb_merged
airbnb_merged.dropna(inplace = True)

In [50]:
airbnb_merged.isna().sum()

listing_id         0
price              0
nbhood_full        0
price_per_month    0
description        0
room_type          0
host_name          0
last_review        0
dtype: int64

In [51]:
# Check if there are any duplicate values
print("There are {} duplicates in the DataFrame.".format(airbnb_merged.duplicated().sum()))

There are 0 duplicates in the DataFrame.


## 8. Analyzing listing prices by NYC borough
<p>Now we have combined all data into a single DataFrame, we will turn our attention to understanding the difference in listing prices between <a href="https://en.wikipedia.org/wiki/Boroughs_of_New_York_City">New York City boroughs</a>. 
We can currently see boroughs listed as the first part of a string within the <code>nbhood_full</code> column, e.g., </p>
<pre><code>Manhattan, Midtown
Brooklyn, Clinton Hill
Manhattan, Murray Hill
Manhattan, Hell's Kitchen
Manhattan, Chinatown</code></pre>
<p></p>
<p>We will therefore need to extract this information from the string and store in a new column, <code>borough</code>, for analysis.</p>

In [53]:
airbnb_merged.head()

listing_id  price                nbhood_full  price_per_month  \
0        2595  225.0         Manhattan, Midtown      6843.750000   
1        3831   89.0     Brooklyn, Clinton Hill      2707.083333   
2        5099  200.0     Manhattan, Murray Hill      6083.333333   
3        5178   79.0  Manhattan, Hell's Kitchen      2402.916667   
4        5238  150.0       Manhattan, Chinatown      4562.500000   

                                 description        room_type    host_name  \
0                      Skylit Midtown Castle  entire home/apt     Jennifer   
1            Cozy Entire Floor of Brownstone  entire home/apt  LisaRoxanne   
2  Large Cozy 1 BR Apartment In Midtown East  entire home/apt        Chris   
3            Large Furnished Room Near B'way     private room     Shunichi   
4         Cute & Cozy Lower East Side 1 bdrm  entire home/apt          Ben   

  last_review  
0  2019-05-21  
1  2019-07-05  
2  2019-06-22  
3  2019-06-24  
4  2019-06-09

In [55]:
# Extract information from the nbhood_full column and store as a new column, borough
airbnb_merged["borough"] = airbnb_merged["nbhood_full"].str.partition(",")[0]
airbnb_merged.head()

listing_id  price                nbhood_full  price_per_month  \
0        2595  225.0         Manhattan, Midtown      6843.750000   
1        3831   89.0     Brooklyn, Clinton Hill      2707.083333   
2        5099  200.0     Manhattan, Murray Hill      6083.333333   
3        5178   79.0  Manhattan, Hell's Kitchen      2402.916667   
4        5238  150.0       Manhattan, Chinatown      4562.500000   

                                 description        room_type    host_name  \
0                      Skylit Midtown Castle  entire home/apt     Jennifer   
1            Cozy Entire Floor of Brownstone  entire home/apt  LisaRoxanne   
2  Large Cozy 1 BR Apartment In Midtown East  entire home/apt        Chris   
3            Large Furnished Room Near B'way     private room     Shunichi   
4         Cute & Cozy Lower East Side 1 bdrm  entire home/apt          Ben   

  last_review    borough  
0  2019-05-21  Manhattan  
1  2019-07-05   Brooklyn  
2  2019-06-22  Manhattan  
3  2019-06-24  Manhattan  
4  2019-06-09  Manhattan

In [57]:
# Group by borough and calculate summary statistics
boroughs = airbnb_merged.groupby("borough")["price"].agg(["sum", "mean", "median", "count"])
boroughs

sum        mean  median  count
borough                                            
Bronx            55156.0   79.247126    65.0    696
Brooklyn       1275250.0  122.021816    95.0  10451
Manhattan      1898417.0  184.044304   149.0  10315
Queens          320715.0   92.826339    70.0   3455
Staten Island    22974.0   86.044944    71.0    267

In [59]:
# Round boroughs to 2 decimal places, and sort by mean in descending order
boroughs = boroughs.round(2).sort_values("mean", ascending= False)
boroughs

sum    mean  median  count
borough                                        
Manhattan      1898417.0  184.04   149.0  10315
Brooklyn       1275250.0  122.02    95.0  10451
Queens          320715.0   92.83    70.0   3455
Staten Island    22974.0   86.04    71.0    267
Bronx            55156.0   79.25    65.0    696

In [60]:
# Print boroughs
print(boroughs)

                     sum    mean  median  count
borough                                        
Manhattan      1898417.0  184.04   149.0  10315
Brooklyn       1275250.0  122.02    95.0  10451
Queens          320715.0   92.83    70.0   3455
Staten Island    22974.0   86.04    71.0    267
Bronx            55156.0   79.25    65.0    696


## 9. Price range by borough
<p>The above output gives us a summary of prices for listings across the 5 boroughs. In this final task we would like to categorize listings based on whether they fall into specific price ranges, and view this by borough. 
<br><br>
We can do this using percentiles and labels to create a new column, <code>price_range</code>, in the DataFrame.
Once we have created the labels, we can then group the data and count frequencies for listings in each price range by borough.
<br><br>
We will assign the following categories and price ranges:</p>
<table>
<thead>
<tr>
<th>label</th>
<th>price</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>Budget</code></td>
<td>\$0-69</td>
</tr>
<tr>
<td><code>Average</code></td>
<td>\$70-175</td>
</tr>
<tr>
<td><code>Expensive</code></td>
<td>\$176-350</td>
</tr>
<tr>
<td><code>Extravagant</code></td>
<td>&gt; \$350</td>
</tr>
</tbody>
</table>

In [61]:
# Create labels for the price range, label_names
label_names = ["Budget", "Average", "Expensive", "Extravagant"]

In [62]:
# Create the label ranges, ranges
ranges = [0, 69, 175, 350, np.inf] #np.inf is infinity

In [64]:
# Insert new column, price_range, into DataFrame
airbnb_merged["price_range"] = pd.cut(airbnb_merged["price"], bins= ranges, labels= label_names)
airbnb_merged["price_range"].head()

0    Expensive
1      Average
2    Expensive
3      Average
4      Average
Name: price_range, dtype: category
Categories (4, object): [Budget < Average < Expensive < Extravagant]

In [66]:
# Calculate borough and price_range frequencies, prices_by_borough
prices_by_borough = airbnb_merged.groupby(["borough", "price_range"])["price_range"].count()
print(prices_by_borough)

borough        price_range
Bronx          Budget          381
               Average         285
               Expensive        25
               Extravagant       5
Brooklyn       Budget         3194
               Average        5532
               Expensive      1466
               Extravagant     259
Manhattan      Budget         1148
               Average        5285
               Expensive      3072
               Extravagant     810
Queens         Budget         1631
               Average        1505
               Expensive       291
               Extravagant      28
Staten Island  Budget          124
               Average         123
               Expensive        20
Name: price_range, dtype: int64
